### PR TITLE
feat(wallet): Choose Quote Provider

### PR DESCRIPTION
**Pull request by Douglashdaniel**
## Description 

Introduces a `Choose quote provider` modal on the `Swap` and `Bridge` screens

Resolves <https://github.com/brave/brave-browser/issues/39472>

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Switching Quote Providers:
1. Open the `Wallet` and navigate to the `Swap` screen
2. Select some tokens to swap and get a quote
3. Open the `Choose quote provider` modal and change providers
4. It should instantly fetch a quote for from that provider.

https://github.com/brave/brave-core/assets/40611140/f78df8d4-9372-442e-af80-bfe2cf411bfe

## Test Available Quote Providers:
1. Open the `Wallet` and navigate to the `Swap` screen
2. Select an `Ethereum` token to swap
3. Open the `Choose quote provider` modal
4. `Jupiter` should be grayed out and `Not available`
5. Change the `Ethereum` token to a `Solana` token
6. Open the `Choose quote provider` modal
7. `Ox` should be grayed out and `Not available`
8. Navigate to the `Bridge` screen and open the `Choose quote provider` modal
9. `Ox` and `Jupiter` should be grayed out and `Not available`

https://github.com/brave/brave-core/assets/40611140/b05c11bd-54bc-4416-ab8c-ebe1a70d7bbb

## Test Unsupported Provider Error:
1. Open the `Wallet` and navigate to the `Swap` screen
2. Open the `Choose quote provider` modal and select `0x`
3. Now select a `Solana` token to swap
4. You should see the error message to `Change provider`

https://github.com/brave/brave-core/assets/40611140/9f979105-2087-4901-88cb-a5caa268a4d8

### CODE REVIEW

File: components/brave_wallet_ui/constants/types.ts [hunk 1]
```diff
@@ -1057,3 +1057,21 @@ export type RewardsExternalWallet = Pick<
 > & {
   status: WalletStatus
 }
+
+type SwapProviderNameType = {
+  [provider: BraveWallet.SwapProvider]: string
+}
```
Comment (josheleonard) [inline components/brave_wallet_ui/constants/types.ts:1063]: Do we need this type? It resolves to ```{ [provider: number]: string }``` which is not very strict

File: components/brave_wallet_ui/page/screens/swap/components/swap/swap_providers/swap_providers.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  // Selectors
+  const isPanel = useSafeUISelector(UISelectors.isPanel)
+
+  // State
+  const [userSelectedSwapProvider, setUserSelectedSwapProvider] =
+    React.useState<BraveWallet.SwapProvider>(selectedProvider)
+
+  // Memos
+  const providerList = React.useMemo(() => {
+    return SupportedSwapProviders.sort(
```
Comment (josheleonard) [inline components/brave_wallet_ui/page/screens/swap/components/swap/swap_providers/swap_providers.tsx:60]: Let's not mutate the original array.

```suggestion
    return [...SupportedSwapProviders].sort(
```

File: components/brave_wallet_ui/page/screens/swap/components/swap/swap_providers/swap_providers.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      >
+        {providerList.map((provider) => (
+          <RadioButton
+            onChange={() => setUserSelectedSwapProvider(provider)}
+            name='providers'
+            value={SwapProviderNameMapping[provider]}
+            currentValue={SwapProviderNameMapping[userSelectedSwapProvider]}
+            key={SwapProviderNameMapping[provider]}
+            isSelected={userSelectedSwapProvider === provider}
+            isDisabled={!availableProvidersForSwap.includes(provider)}
```
Comment (josheleonard) [inline components/brave_wallet_ui/page/screens/swap/components/swap/swap_providers/swap_providers.tsx:100]: perf: let's compute this once per map:
```const isAvailableForSwap = availableProvidersForSwap.includes(provider)```

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts [hunk 4]
```diff
@@ -848,8 +852,39 @@ export const useSwap = () => {
...
+    if (isBridge) {
+      return defaultProviders
+    }
+    // Returns all providers if no from token has been selected yet.
+    if (!fromToken) {
+      return [
+        ...defaultProviders,
+        BraveWallet.SwapProvider.kZeroEx,
+        BraveWallet.SwapProvider.kJupiter
+      ]
```
Comment (josheleonard) [inline components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts:871]: Perf: can be defined outside of the hook

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts [hunk 5]
```diff
@@ -848,8 +852,39 @@ export const useSwap = () => {
...
+    }
+    // Returns default providers and Jupiter for SOL swaps.
+    if (fromToken && fromToken.coin === BraveWallet.CoinType.SOL) {
+      return [...defaultProviders, BraveWallet.SwapProvider.kJupiter]
```
Comment (josheleonard) [inline components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts:875]: Perf: can be defined outside of the hook

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts [hunk 6]
```diff
@@ -848,8 +852,39 @@ export const useSwap = () => {
...
+    }
+    // Returns default providers and Ox for ETH swaps.
+    return [...defaultProviders, BraveWallet.SwapProvider.kZeroEx]
```
Comment (josheleonard) [inline components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts:878]: Perf: can be defined outside of the hook

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts [hunk 7]
```diff
@@ -848,8 +852,39 @@ export const useSwap = () => {
...
+  }, [isBridge, fromToken])
+
   const swapValidationError: SwapValidationErrorType | undefined =
     useMemo(() => {
+      // Display error is provider is not supported for a swap.
```
[the remark this answers is not in the document]
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts]: Thank! Fixed this!

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts [hunk 8]
```diff
@@ -848,8 +852,39 @@ export const useSwap = () => {
     return Amount.zero()
   }, [quoteOptions])
 
+  const availableProvidersForSwap = useMemo(() => {
+    // Auto and Li.Fi are always supported.
+    const defaultProviders = [
+      BraveWallet.SwapProvider.kAuto,
+      BraveWallet.SwapProvider.kLiFi
+    ]
```
[the remark this answers is not in the document]
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts]: Moved!

File: components/brave_wallet_ui/constants/types.ts (hunk 1 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/constants/types.ts]: Removed!

File: components/brave_wallet_ui/page/screens/swap/components/swap/swap_providers/swap_providers.tsx (hunk 2 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/swap/components/swap/swap_providers/swap_providers.tsx]: Done!

File: components/brave_wallet_ui/page/screens/swap/components/swap/swap_providers/swap_providers.tsx (hunk 3 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/swap/components/swap/swap_providers/swap_providers.tsx]: Done!

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts (hunk 4 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts]: Moved!

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts (hunk 5 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts]: Moved!

File: components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts (hunk 6 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts]: Moved!

**Review by josheleonard — APPROVED** [review]
++

**Review by nvonpentz — APPROVED** [review]

**Review by mkarolin — APPROVED** [review]
`strings`++

**Pull request merged**